src/dwarf/debug_info.rs

```diff
@@ -221,336 +230,140 @@ pub enum AttrValue<'a> {
 ///
 /// Return AttrValue and the number of bytes it takes.
 fn extract_attr_value(
-    data: &[u8],
+    mut data: &[u8],
     form: u8,
     dwarf_sz: usize,
     addr_sz: usize,
-) -> Option<(AttrValue, usize)> {
+) -> Option<(AttrValue<'_>, usize)> {
     match form {
         DW_FORM_addr => {
             if addr_sz == 0x4 {
-                if 4 <= data.len() {
-                    Some((AttrValue::Unsigned(decode_uword(data) as u64), 4))
-                } else {
-                    None
-                }
-            } else if 8 <= data.len() {
-                Some((AttrValue::Unsigned(decode_udword(data)), 8))
+                Some((AttrValue::Unsigned(data.read_u32()? as u64), 4))
             } else {
-                None
+                Some((AttrValue::Unsigned(data.read_u64()?), 8))
             }
         }
         DW_FORM_block2 => {
-            if 2 <= data.len() {
-                let bytes = decode_uhalf(data);
-                let fullsize = bytes as usize + 2;
-                if fullsize <= data.len() {
-                    Some((AttrValue::Bytes(&data[2..fullsize]), fullsize))
-                } else {
-                    None
-                }
-            } else {
-                None
-            }
+            let bytes = data.read_u16()?;
+            let fullsize = bytes as usize + 2;
+            Some((AttrValue::Bytes(data.read_slice(bytes.into())?), fullsize))
         }
         DW_FORM_block4 => {
-            if 2 <= data.len() {
-                let bytes = decode_uword(data);
-                let fullsize = bytes as usize + 2;
-                if fullsize <= data.len() {
-                    Some((AttrValue::Bytes(&data[2..fullsize]), fullsize))
-                } else {
-                    None
-                }
-            } else {
-                None
-            }
-        }
-        DW_FORM_data2 => {
-            if 2 <= data.len() {
-                let v = decode_uhalf(data);
-                Some((AttrValue::Unsigned(v as u64), 2))
-            } else {
-                None
-            }
-        }
-        DW_FORM_data4 => {
-            if 4 <= data.len() {
-                let v = decode_uword(data);
-                Some((AttrValue::Unsigned(v as u64), 4))
-            } else {
-                None
-            }
-        }
-        DW_FORM_data8 => {
-            if 8 <= data.len() {
-                let v = decode_udword(data);
-                Some((AttrValue::Unsigned(v), 8))
-            } else {
-                None
-            }
+            let bytes = data.read_u32()?;
+            let fullsize = bytes as usize + 2;
```

Comment:
`+ 4`? we are adding prefix size here, right?

Reply:
though I see now that with the next commits it becomes irrelevant

Reply:
Yes, that is https://github.com/libbpf/blazesym/issues/55 and existing code. As you said it is addressed by the last commit in the series.